src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/SignatureHelp/RemoteSignatureHelpService.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Razor.DocumentMapping;
+using Microsoft.CodeAnalysis.Razor.Remote;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.ServiceHub.Framework;
+using Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor;
+
+using SignatureHelp = Roslyn.LanguageServer.Protocol.SignatureHelp;
+
+internal sealed class RemoteSignatureHelpService(
+    IServiceBroker serviceBroker,
+    DocumentSnapshotFactory documentSnapshotFactory,
+    IFilePathService filePathService,
+    IRazorDocumentMappingService documentMappingService)
+    : RazorDocumentServiceBase(serviceBroker, documentSnapshotFactory), IRemoteSignatureHelpService
+{
+    private readonly IFilePathService _filePathService = filePathService;
+    private readonly IRazorDocumentMappingService _documentMappingService = documentMappingService;
+
+    public ValueTask<SignatureHelp?> GetSignatureHelpAsync(JsonSerializableRazorPinnedSolutionInfoWrapper solutionInfo, JsonSerializableDocumentId documentId, Position position, CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetSignatureHelpsAsync(context, position, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<SignatureHelp?> GetSignatureHelpsAsync(RemoteDocumentContext context, Position position, CancellationToken cancellationToken)
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var linePosition = new LinePosition(position.Line, position.Character);
+        var absoluteIndex = linePosition.GetRequiredAbsoluteIndex(codeDocument.Source.Text, logger: null);
+
+        var generatedDocument = await context.GetGeneratedDocumentAsync(_filePathService, cancellationToken).ConfigureAwait(false);
+
+        if (_documentMappingService.TryMapToGeneratedDocumentPosition(codeDocument.GetCSharpDocument(), absoluteIndex, out var mappedPosition, out _))
+        {
+            return await ExternalAccess.Razor.Cohost.Handlers.SignatureHelp.GetSignatureHelpAsync(generatedDocument, mappedPosition, supportsVisualStudioExtensions: true, cancellationToken).ConfigureAwait(false);
```

Comment:
>ExternalAccess.Razor.Cohost.Handlers.SignatureHelp [](http://example.com/codeflow?start=25&length=50)

nit: intentionally using namespace-qualified type name here?

Reply:
The word "SignatureHelp" refers a namespace in Razor, a namespace in Roslyn, a type name in Roslyn, and a type name in Microsoft.VisualStudio.LanguageServer.Protocol. The fact that this whole thing compiles at all is a magically juggling act of using aliases. Rest assured I did the bare minimum 😁

Reply:
Fair enough :)

Reply:
I wonder if finding a naming convention for our Roslyn entry points here would help. I named them after the handler they expose functionality from, but it occurs to me thats almost guaranteed to conflict with something. I don't like the idea of `SignatureHelpHelper`, but if anyone has naming suggestions, feel free to shout them out.